transports/onion: Add dial-only implementation of `Transport`

umgefahren opened the pull request:
# Description
<!-- Please write a summary of your changes and why you made them.-->
Adding a libp2p transport built on top of the newly released [Arti](https://gitlab.torproject.org/tpo/core/arti). Currently only dialing is supported, since Onion Services are not supported by Arti yet and it will likely take some time until it's supported, since it's not funded yet.

The implementation should be considered as a first step towards a stable Tor based transport.

## Links to any relevant issues
<!-- Reference any related issues.-->
This implementation follows the discussion in #708 and should close this issue.

## Open Questions
<!-- Unresolved questions, if any. -->
~~I added libp2p-onion to the default features. However this might be a bad idea, since arti has a lot of dependencies and therefore increases the build time of libp2p greatly.~~ (There are no default features anymore)

Since ~~there is no documentation yet and~~ it's very untested (because of the transports nature automated testing is hard) ~~libp2p-onion maybe shouldn't be integrated in libp2p at all now~~.

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

umgefahren committed:
Initial commit on Onion transport for libp2p

umgefahren committed:
Added tests, comments and mate implementation somewhat usesable

umgefahren committed:
simplified, streamlined and added an example

umgefahren committed:
simplified, streamlined and added an example

umgefahren committed:
Added some traits

umgefahren committed:
Added cargo fmt changes

umgefahren commented on examples/ping-onion.rs:
```diff
@@ -0,0 +1,121 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
```
I'm not quite sure this copyright notice is correct.

mxinden replied:
You would replace the top line with your name:

```
Copyright 2022 XXX
```

Alternatively you can use:

```
Copyright 2022 Protocol Labs
```

umgefahren replied:
Thanks for the suggestion. I went for the first option :)

umgefahren committed:
fixed typo

umgefahren reviewed (commented):
Documentation is missing, but I will add that, once the code itself is approved.

thomaseizinger commented on examples/ping-onion.rs:
```diff
@@ -0,0 +1,122 @@
+        let dns_tcp = dns::DnsConfig::system(tcp::TcpTransport::new(
+            tcp::GenTcpConfig::new().nodelay(true),
+        ))
+        .await?;
```
Why do we need this to showcase a dial-only onion transport?

umgefahren replied:
We actually don't. We could just connect to the normal examples/ping, I just thought it would be easier that way.

thomaseizinger commented on examples/ping-onion.rs:
```diff
@@ -0,0 +1,122 @@
+
+    // Tell the swarm to listen on all interfaces and a random, OS-assigned
+    // port.
+    swarm.listen_on("/ip4/0.0.0.0/tcp/0".parse()?)?;
```
That seems unnecessary?

umgefahren replied:
I think it's necessary in order to connect to the node.

thomaseizinger replied:
We'll have to think of a different example to actually showcase Tor. The mix of TCP, DNS and Tor is quite confusing here. I'd prefer something that just uses Tor.

If we want to have a separate listening service, I think it would be good to split this into two, where the listening service doesn't know anything about Tor and the dialing service only uses Tor.

thomaseizinger commented on transports/onion/src/address.rs:
```diff
@@ -0,0 +1,34 @@
+) -> Result<TorAddr, TorAddrError> {
+    let socket_addr = try_extract_socket_addr(multiaddr)?;
+    TorAddr::dangerously_from(socket_addr)
+}
```
From reaching the docs of `dangerously_from`, I think we should only be supporting `/dns` multiaddr because we can't guarantee where the IP comes from and whether it will have leaked the user's IP through a DNS resolution.

umgefahren replied:
I think we should allow connections from IP addresses but also warn in the doc. I will implement the changes.

umgefahren replied:
Resolved in latest commit.

thomaseizinger commented on transports/onion/tests/mod.rs:
```diff
@@ -0,0 +1,88 @@
+};
+use libp2p_onion::OnionClient;
+
+#[cfg(feature = "async-std")]
```
Instead of this, you can define the test in `Cargo.toml` explicitly and say it requires certain features. See here for example: https://github.com/Restioson/xtra/blob/b538a5a41130fbfeb3e14eb816e1010f9167d17a/Cargo.toml#L114-L116

umgefahren replied:
Resolved in latest commit.

thomaseizinger commented on transports/onion/tests/mod.rs:
```diff
@@ -0,0 +1,88 @@
+        let dns_tcp = dns::DnsConfig::system(tcp::TcpTransport::new(
+            tcp::GenTcpConfig::new().nodelay(true),
+        ))
+        .await?;
```
Let's not use this transport here.

umgefahren replied:
Resolved in latest commit.

thomaseizinger commented on transports/onion/tests/mod.rs:
```diff
@@ -0,0 +1,88 @@
+            _ => {}
+        }
+    }
+}
```
This doesn't test anything about the onion transport itself but just the embedded TCP transport.

thomaseizinger replied:
I am not sure if we can provide _good_ tests here. In the end, all the heavy lifting is done by arti which we can trust in. All we are doing is a bit of plumbing and that is mostly secured by the type system so I am not sure we need tests for this crate on this level.

Obviously, a manual test might not hurt :)

umgefahren replied:
I will delete the test, since it seems such a bad idea.

umgefahren replied:
Resolved in latest commit.

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,123 @@
+    // client is in an Arc, because wihtout it the Transport::Dial method can't be implemented,
+    // due to lifetime issues. With the, eventual, stabilization of static async traits this issue
+    // will be resolved.
+    client: Arc<TorClient<PreferredRuntime>>,
```
Although I think the Tor people did a great job with this abstraction, I the way he handled this in libp2p so far is that we expose multiple transports, one for each runtime.

See `libp2p-tcp` as an example.

umgefahren replied:
Resolved in latest commit.

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,123 @@
+        self: std::pin::Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        core::task::Poll::Ready(Err(OnionError::OnionServiceUnimplemented))
```
You could also panic here because it will never be called, that would remove an error variant from the public API.

umgefahren replied:
Resolved in latest commit.

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,123 @@
+    client: Arc<TorClient<PreferredRuntime>>,
+}
+
+pub type OnionBuilder = TorClientBuilder<PreferredRuntime>;
```
This is leaking the `arti_client` dependency to our consumers, meaning it affects semver when we upgrade the `arti_client` dependency.

I think that is okay for now.

thomaseizinger commented on transports/onion/src/provider.rs:
```diff
@@ -0,0 +1,132 @@
+
+#[derive(Debug)]
+pub struct OnionStream {
+    inner: DataStream,
```
Can we just expose `DataStream` directly? Or does that not work because something further down the stack requires `futures::Async{Read,Write}` even with the `tokio` feature`?

thomaseizinger replied:
I am not happy with these `cfg`s in here. Will need to give this more thought :)

umgefahren replied:
> Can we just expose `DataStream` directly? Or does that not work because something further down the stack requires `futures::Async{Read,Write}` even with the `tokio` feature`?

I'm actually not quite sure. Although this part is, along with the cfg's, ported from libp2p-tcp.

thomaseizinger replied:
> > Can we just expose `DataStream` directly? Or does that not work because something further down the stack requires `futures::Async{Read,Write}` even with the `tokio` feature`?
> 
> I'm actually not quite sure. Although this part is, along with the cfg's, ported from libp2p-tcp.

Oh interesting, I'll need to take another look at what is going on there then!

thomaseizinger replied:
I had a look at what we currently do in `libp2p-tcp` and I think it is worth doing the same thing here:

1. For `async-io`, we can directly expose `DataStream` because it implements the traits directly, I don't think it is necessary to wrap that again.
2. For `tokio`, we can wrap `DataStream` and use the adaptors to implement `futures::Async{Read,Write}`.

That way, it is clear which types are exposed for which features and enabling all of them is not a problem and does not require any boolean operations on the feature set.

To realize this, we may a `Provider` abstraction, similar to what we do in `libp2p-tcp`. It should be fairly simple though because we only need to adapt what we do with the return value from `TorClient::connect`.

thomaseizinger reviewed (commented):
Did an initial pass, thanks for tackling this!

umgefahren committed:
implemented suggestions

umgefahren committed:
implemented suggestions

thomaseizinger commented on transports/onion/src/address.rs:
```diff
@@ -32,3 +32,49 @@ pub(super) fn dangerous_extract_tor_address(
+    };
+}
+
+pub(super) fn safe_extract_tor_address(multiaddr: &Multiaddr) -> Result<TorAddr, TorAddrError> {
```
Can we have some unit tests for this function please?

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -22,37 +23,95 @@ pub enum OnionError {
     #[inline]
-    pub fn builder() -> OnionBuilder {
-        TorClient::builder()
+    pub fn from_builder(builder: OnionBuilder<R>, safe_mode: bool) -> Result<Self, OnionError> {
```
Instead of a bool, can we use an enum here? That would be much more self-documenting. I am thinking something like:

```rust
pub enum AddressConversion {
    DnsOnly,
    IpAndDns
}
```

The rustdoc on this enum can then link to Tor's documentation of `DangerouslyIntoTorAddress` etc which will explain the privacy caveats.

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -22,37 +23,95 @@ pub enum OnionError {
+            .expect("Couldn't get the current tokio rustls runtime");
+        TorClient::with_runtime(runtime)
+    }
+}
```
Can we have some convenience constructors here that do everything in one go with sensible defaults?

- Bootstrap the client
- Enable safe mode

thomaseizinger commented on transports/onion/src/address.rs:
```diff
@@ -0,0 +1,80 @@
+    Ok(SocketAddr::new(ip, tcp_port))
+}
+
+pub(super) fn dangerous_extract_tor_address(
```
`address` is a private module so this can be simplified to `pub fn`,

thomaseizinger commented on transports/onion/src/address.rs:
```diff
@@ -0,0 +1,80 @@
+use arti_client::{IntoTorAddr, TorAddr, TorAddrError};
+use libp2p_core::{multiaddr::Protocol, Multiaddr};
+
+fn try_extract_socket_addr(multiaddr: &Multiaddr) -> Result<SocketAddr, TorAddrError> {
```
Can we order the functions in this file according to their visibility / usage please?

Public functions first, followed by utilities in the order they are used as you read code from top to bottom. Thanks!

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,193 @@
+        libp2p_core::transport::ListenerId,
+        libp2p_core::transport::TransportError<Self::Error>,
+    > {
+        // although this address might be supported, this is returned in order to not provoke an
+        // error when trying to listen on this transport.
```
This comment is not quite true I think because at the moment, we simply don't support listening on anything because we haven't even defined what listening on an onion transport is meant to do.

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,193 @@
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<libp2p_core::transport::TransportEvent<Self::ListenerUpgrade, Self::Error>>
+    {
+        // pending is returned here, because this won't panic an OrTransport.
```
```suggestion
        // pending is returned here because this transport doesn't support listening
```

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,193 @@
+        Err(TransportError::MultiaddrNotSupported(addr))
+    }
+
+    /// Always returns false
```
```suggestion
```

Don't think we need that :)

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    /// always returns pending
+    fn poll(
```
Same here.

```suggestion
    fn dial_as_listener(
        &mut self,
        addr: Multiaddr,
    ) -> Result<Self::Dial, TransportError<Self::Error>> {
        self.dial(addr)
    }

    fn address_translation(&self, _listen: &Multiaddr, _observed: &Multiaddr) -> Option<Multiaddr> {
        None
    }

    fn poll(
```

thomaseizinger commented on transports/onion/src/address.rs:
```diff
@@ -0,0 +1,80 @@
+            }
+            _ => {}
+        }
+    }
```
I am not sure if looping towards the end of the address is the best approach here.

I'd suggest we instead pop off two items from the iterator, ignore the rest and match the first one for any of the DNS variants and the second one on `Tcp`. If they are anything else, bail with `MultiAddrNotSuppored` (this is currently done wrong in the `dial` implementation which means the next transport will not get selected in a chain of transports).

This should allow you to not use mutability but just directly assign the variables.

thomaseizinger commented on transports/onion/src/address.rs:
```diff
@@ -0,0 +1,80 @@
+            }
+            Protocol::Dnsaddr(s) => {
+                dns_addr = Some(s);
+            }
```
`/dnsaddr` is a specific protocol that requires resolution of TXT records which won't work with Tor so we shouldn't be using this here!

umgefahren replied:
Oh I see

thomaseizinger commented on Cargo.toml:
```diff
@@ -68,6 +71,11 @@ yamux = ["dep:libp2p-yamux"]
 secp256k1 = ["libp2p-core/secp256k1"]
 rsa = ["libp2p-core/rsa"]
 serde = ["libp2p-core/serde", "libp2p-kad?/serde", "libp2p-gossipsub?/serde"]
+onion = ["dep:libp2p-onion"]
```
Is this feature-flag by itself useful?

umgefahren replied:
No. I will remove it.

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,193 @@
+            safe_extract_tor_address(&addr)
+        }
+        .map_err(OnionError::from)
+        .map_err(TransportError::Other)?;
```
Here we should return a `TransportError::MultiaddrNotSupported` in case we can't dial the address.

umgefahren replied:
Resolved in latest commit.

thomaseizinger reviewed (commented):
Thanks! I've left a few more comments with a mix of nits and other feedback :)

umgefahren committed:
Merge branch 'master' into add-onion

umgefahren committed:
Addressed latest issues

umgefahren committed:
Merge branch 'add-onion' of github.com:umgefahren/rust-libp2p into add-onion

umgefahren committed:
fixed test failure

umgefahren committed:
Ran cargo fmt

thomaseizinger commented on examples/ping-onion.rs:
```diff
@@ -0,0 +1,119 @@
+        .expect("Signing libp2p-noise static DH keypair failed.");
+    Ok(transport
+        .upgrade(upgrade::Version::V1)
+        .authenticate(noise::NoiseConfig::xx(noise_keys).into_authenticated())
```
A [recent PR](https://github.com/libp2p/rust-libp2p/pull/2887) simplified the construction of a noise authentication layer. Please use that here too for consistency.

thomaseizinger commented on transports/onion/src/address.rs:
```diff
@@ -0,0 +1,107 @@
+                let new_addr_res = safe_extract_tor_address(&mut new);
+                assert!(new_addr_res.is_ok());
+            }
+        }
```
What are we testing here?

umgefahren replied:
Possibly uncovered combinations and longer multiaddresses. I can delete it though.

thomaseizinger replied:
Despite multi-addresses allowing for a lot of nesting, I am not sure an address in the form of `/dns4/domain.tld/tcp/8080/dns4/other.tld` makes any sense :/

thomaseizinger commented on transports/onion/src/address.rs:
```diff
@@ -0,0 +1,107 @@
+        let dns_6_test_addr = multiaddr!(Dns6(Cow::Borrowed("dns.ip6.tld")), Tcp(12u16));
+        let dns_6_test_addr_res = safe_extract_tor_address(&mut dns_6_test_addr.clone());
+        assert!(dns_6_test_addr_res.is_ok());
+        assert!(dns_6_test_addr_res.unwrap().is_ip_address().not());
```
I think this would be clearer if we follow the schema of:

- arrange (construct a list of multiaddresses)
- act (try and convert them into tor addresses)
- assert (that they are all `Ok`)

Currently, there is a lot of text which makes it hard to see, where the input parameters to the test are and what is actually being asserted.

thomaseizinger commented on transports/onion/src/address.rs:
```diff
@@ -0,0 +1,107 @@
+
+    #[test]
+    fn extract_correct_address() {
+        let dns_test_addr = multiaddr!(Dns(Cow::Borrowed("ip.tld")), Tcp(10u16));
```
You may find that constructing the mutliaddress from a string and parsing it is easier to read, i.e. this line could be written as:

```suggestion
        let dns_test_addr = "/dns/ip.tld/tcp/10".parse::<Multiaddr>().unwrap();
```

umgefahren replied:
I agree, I just always wanted to try that macro 🙃

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,207 @@
+    /// uses only dns for address resolution
+    #[default]
+    DnsOnly,
+    /// uses ip and dns for addresses
```
Please link to the docs `arti` docs of `TorAddr::dangerously_from` here.

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,207 @@
+
+macro_rules! default_constructor {
+    () => {
+        pub async fn default() -> Result<Self, OnionError> {
```
Nit: I think people associate `default` with a very particular signature, the one from the `Default` trait. Maybe this is better as `new` or perhaps `bootstrapped`?

umgefahren replied:
`new` would fit very good in rust conventions, however `bootstrapped` would be better. I chose `default` exactly because it resembles the `Default` trait, which I would chose, if it would allow for async constructors.

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,207 @@
+    // due to lifetime issues. With the, eventual, stabilization of static async traits this issue
+    // will be resolved.
+    client: Arc<TorClient<R>>,
+    pub conversion_mode: AddressConversion,
```
Instead of making this public, what do you think about a `with_address_conversion` function? That would allow people to write:

```rust
let client = OnionClient::default().await?.with_address_conversion(AddressConversion::IpAndDns);
```

umgefahren replied:
sounds good and convenient, however I don't think it conflicts with making conversion mode public. There might be reasons why someone would like to change the address conversion mode later after construction.

thomaseizinger replied:
The `with_` function does allow for that though, right? The only thing that is not possible is reading the mode but given that it is static, I am not sure why one would need that.

Another though I just had is that we could embed it in the type signature too. That could allow client code to require a "safe" OnionTransport in a type-safe way. Might be overkill though!

umgefahren replied:
> The with_ function does allow for that though, right? The only thing that is not possible is reading the mode but given that it is static, I am not sure why one would need that.

For sure, I just considered it less natural. I will however remove the `pub` modifier if thats whats necessary.

> Another though I just had is that we could embed it in the type signature too. That could allow client code to require a "safe" OnionTransport in a type-safe way. Might be overkill though!
It might be possible, but I think it's not necessary. The API is pretty obfuscated now anyway and I don't see a real benefit.

thomaseizinger replied:
I think this is a good compromise. Having the field pub allows users to read it and the `with` function makes it possible to chain it onto an existing instance.

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,207 @@
+#[derive(Debug, thiserror::Error)]
+pub enum OnionError {
+    #[error("error during address translation")]
+    AddrErr(#[from] TorAddrError),
```
Is this variant used anywhere, now that we convert that to `MultiaddrNotSupported`?

thomaseizinger replied:
I think this may allow us to just use `arti_client::Error` and not need `OnionError` at all?

thomaseizinger commented on examples/ping-onion.rs:
````diff
@@ -0,0 +1,119 @@
+//! ```
+//!
+//! The two nodes establish a connection, negotiate the ping protocol
+//! and begin pinging each other.
````
This example text needs to be updated.

For this to work with the original ping example, the user will need to listen on an external interface and open a port so that their IP is reachable from the internet and thus the Tor network.

Then they can run the ping-onion example against that IP.

thomaseizinger commented on examples/ping-onion.rs:
```diff
@@ -0,0 +1,119 @@
+        onion.bootstrap().await?;
+        println!("bootstrapped!");
+        onion
+    };
```
Could make use of our new constructor here.

thomaseizinger commented on transports/onion/src/address.rs:
```diff
@@ -0,0 +1,107 @@
+use libp2p_core::{multiaddr::Protocol, Multiaddr};
+
+pub fn dangerous_extract_tor_address(multiaddr: &Multiaddr) -> Result<TorAddr, TorAddrError> {
+    let socket_addr = try_extract_socket_addr(&mut multiaddr.clone())?;
```
If we end up cloning it, we may as well require ownership of the address directly.

thomaseizinger commented on transports/onion/src/address.rs:
```diff
@@ -0,0 +1,107 @@
+    TorAddr::dangerously_from(socket_addr)
+}
+
+pub fn safe_extract_tor_address(multiaddr: &mut Multiaddr) -> Result<TorAddr, TorAddrError> {
```
To align the signatures, this should probably take an owned version too.

umgefahren replied:
I could do that but that might be incorrect because, as you said:
> I'd suggest we instead pop off two items from the iterator, ignore the rest and match the first one for any of the DNS variants and the second one on `Tcp`. If they are anything else, bail with `MultiAddrNotSuppored` (this is currently done wrong in the `dial` implementation which means the next transport will not get selected in a chain of transports).

If it's not mutable we don't get the changes to the multiaddress. However I think there is an easy solution.

umgefahren replied:
Nevermind, I don't see an ideal solution. If we want to make the pop operations count i consider `&mut` a way better solution. I tried it with owned versions, however this would require way more clones. If you have a better idea, I would be glad.

thomaseizinger replied:
I've pushed 8c20c301, let me know what you think :)

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,207 @@
+}
+
+#[derive(Clone)]
+pub struct OnionClient<R: Runtime> {
```
```suggestion
pub struct OnionTransport<R: Runtime> {
```

This follows the naming convention from the remaining crates.

thomaseizinger reviewed (commented):
Thanks for the continued work here :blush: 

I left a few more comments. Looking forward to merging this!

umgefahren committed:
Implemented suggested changes

umgefahren committed:
Small module doc

umgefahren committed:
compiles again on wasm32

thomaseizinger commented on examples/ping-onion.rs:
````diff
@@ -0,0 +1,114 @@
+//!
+//! On the second computer run:
+//! ```sh
+//! cargo run --example ping-onion -- /ip4/127.0.0.1/tcp/24915
````
Nit: 127.0.0.1 will never be reachable from the internet :upside_down_face:

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,232 @@
+#![cfg_attr(docsrs, feature(doc_cfg))]
+//! Tor based transport for libp2p. Connect through the Tor network to TCP listeners.
+//!
+//! Main entrypoint of the crate: [OnionTransport]
```
```suggestion
//! Main entrypoint of the crate: [`OnionTransport`]
```

This will make sure rustdoc flags it as a bad link if it is ever renamed.

thomaseizinger commented on examples/ping-onion.rs:
```diff
@@ -0,0 +1,114 @@
+> {
+    use std::time::Duration;
+
+    let transport = onion::OnionAsyncStdNativeTlsTransport::bootstrapped().await?;
```
I believe for this example to work we will need to adjust the `AddressConversion` mode here, otherwise we can dial an IP address.

thomaseizinger committed:
Refactor multiaddress parsing to avoid clones

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,232 @@
+        self.client.bootstrap().await
+    }
+
+    pub fn with_address_conversion(&mut self, conversion_mode: AddressConversion) -> &mut Self {
```
I am never quite sure what the Rust-idiomatic fn signature for builder functions is. I typically go for `self` because it allows usages like:

```rust
let client = OnionTransport::bootstrapped().await?.with_address_conversion(AddressConversion::IpAndDns);
```

Using `&mut self` forces an intermediary variable but I've had people [complain](https://github.com/testcontainers/testcontainers-rs/issues/382) about that in the past. Thoughts?

umgefahren replied:
Yes it's actually quite difficult and there is no really elegant solution in my opinion. Using Self is less flexible and requires assignments on the other hand it's more clean. I preferred `&mut self` because it allows for changing the conversion mode later on, while Self would be only really appropriate during creation. Since we added this to allow conversion during creation however `Self` would be better.

thomaseizinger replied:
Yeah, there isn't really a usecase where you hold on to a `Transport` in libp2p. You construct them and pass them to the `Swarm`, at which point it is moved.

thomaseizinger commented on transports/onion/src/provider.rs:
```diff
@@ -0,0 +1,84 @@
+use arti_client::DataStream;
+use futures::{AsyncRead, AsyncWrite};
+
+pub trait OnionStream: AsyncRead + AsyncWrite + From<DataStream> {}
+
+impl OnionStream for DataStream {}
```
Nice idea!

thomaseizinger commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,232 @@
+
+#[cfg(all(feature = "native-tls", feature = "async-std"))]
+#[cfg_attr(docsrs, doc(cfg(all(feature = "native-tls", feature = "async-std"))))]
+pub type OnionAsyncStdNativeTlsTransport =
```
Nit: I find it a bit odd that we are splitting `Onion` and `Transport` apart here. How about:

```suggestion
pub type AsyncStdNativeTlsOnionTransport =
```

thomaseizinger reviewed (commented):
Thanks!

Let me know what you think of the commit I pushed. A few minor doc fixes, otherwise happy to merge this :)

Would you mind doing a smoke test with the example to check that we aren't shipping something outright broken?

umgefahren reviewed (commented):
Very nice solution :)

umgefahren committed:
Apply suggestions from code review

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

umgefahren committed:
renamed stuff

umgefahren committed:
corrected issues of previous commit an implemented all recently requested changes

umgefahren committed:
formatted

mxinden commented on examples/ping-onion.rs:
```diff
@@ -0,0 +1,117 @@
+// Copyright 2022 Hannes Furmans
```
I would suggest moving this example to `transports/onion/examples`. Not a strong opinion though.

Either way, I think it should be listed on `examples/README.md`.

umgefahren replied:
I think we should leave it there, since all examples are in the examples/ directory and not in the corresponding parts. If we change that for all, we should change it for libp2p-onion too, however I think this is more concise.

thomaseizinger replied:
> I think we should leave it there, since all examples are in the examples/ directory and not in the corresponding parts. If we change that for all, we should change it for libp2p-onion too, however I think this is more concise.

IMO the examples in `/examples` should showcase use of the root `libp2p` crate which usually involves multiple different protocols (identify + kad + ping etc). Most protocols in `/protocols` have their own examples that showcase specific aspects of them.

With this reasoning, the `ping` example shouldn't be there though and neither should `mdns-passive-discovery` and `gossipsub-chat`.

mxinden commented on transports/onion/Cargo.toml:
```diff
@@ -0,0 +1,33 @@
+license = "MIT"
+resolver = "2"
+
+
```
```suggestion
```

Same below.

mxinden commented on transports/onion/Cargo.toml:
```diff
@@ -0,0 +1,33 @@
+[features]
+default = ["async-std", "native-tls"]
+tokio = ["arti-client/tokio", "dep:tokio-crate"]
+async-std = ["arti-client/async-std"]
```
```suggestion
async-std = ["arti-client/async-std", "dep:async-std-crate"]
```

I might be missing something here.

umgefahren replied:
I might be missing something too, but it doesn't seem to change anything.

thomaseizinger replied:
Have you tried running with `--no-default-features --feature async-std`? That should trigger a problem I think.

mxinden commented on transports/onion/Cargo.toml:
```diff
@@ -0,0 +1,33 @@
+native-tls = ["arti-client/native-tls"]
+rustls = ["arti-client/rustls"]
+
+[package.metadata.docs.rs]
+rustdoc-args = ["--cfg", "docsrs"]
+rustc-args = ["--cfg", "docsrs"]
```
What is this needed for?

umgefahren replied:
I can't really try this, however I used this in my other crates in order to enable the display of features in the docs.

thomaseizinger replied:
What this does is tell docs.rs to pass the `docsrs` conditional compilation flag to rustc, meaning any code that is guarded with `cfg(docsrs)` will now be included in the build.

We use this in the `libp2p` root crate to conditionally include the `doc` attribute: https://github.com/libp2p/rust-libp2p/blob/04b07c4d14a0a6ec350fc21261a33d859fc71289/src/lib.rs#L53

We need to do this conditionally because the `#[doc(cfg)]` attribute is [not yet stable](https://github.com/rust-lang/rust/issues/43781). docs.rs runs all their build pipelines on nightly so we can use it there.

Now, interestingly, we don't set this in our `libp2p` crate so all of these `cfg_attr` attributes have been doing nothing all this time :)

In this crate, it is used here: https://github.com/umgefahren/rust-libp2p/blob/6f4bca5bbfd141bc18d9d931c5d28b50a212f0fe/transports/onion/src/provider.rs#L31-L36

And will tell docs.rs to include a little "only available with feature `tokio`" tag next to the item.

thomaseizinger replied:
This is actually something that should be fixed. I'll open an issue about it.

mxinden commented on transports/onion/src/address.rs:
```diff
@@ -0,0 +1,147 @@
+use arti_client::{DangerouslyIntoTorAddr, IntoTorAddr, TorAddr};
```
Missing license header.

umgefahren replied:
Resolved in latest commit.

mxinden commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,233 @@
+#![doc(html_logo_url = "https://libp2p.io/img/logo_small.png")]
```
Missing license header.

umgefahren replied:
Resolved in latest commit.

mxinden commented on transports/onion/src/lib.rs:
````diff
@@ -0,0 +1,233 @@
+//!
+//! ## Example
+//! ```no_run
+//! # use async_std_crate as async_std;
+//! # use libp2p_core::Transport;
+//! # async fn test_func() -> Result<(), Box<dyn std::error::Error>> {
+//! let address = "/dns/www.torproject.org/tcp/1000".parse()?;
+//! let mut transport = libp2p_onion::AsyncStdNativeTlsOnionTransport::bootstrapped().await?;
+//! // we have achieved tor connection
+//! let _conn = transport.dial(address)?.await?;
+//! # Ok(())
+//! # }
+//! # async_std::task::block_on(async { test_func().await.unwrap() });
````
Would you mind fixing the indentation here? As far as I know it will propagate to the rendered HTML.

umgefahren replied:
<img width="979" alt="image" src="https://user-images.githubusercontent.com/55623006/192736528-c92f412c-fd1f-4e5d-84ea-3faed1030fe1.png">

I don't see any indentation issues, if you see some, i will correct them.

mxinden commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,233 @@
+
+#[derive(Clone)]
+pub struct OnionTransport<R: Runtime, S> {
+    // client is in an Arc, because wihtout it the Transport::Dial method can't be implemented,
```
```suggestion
    // client is in an Arc, because without it the [`Transport::dial`] method can't be implemented,
```

mxinden commented on transports/onion/src/lib.rs:
```diff
@@ -0,0 +1,233 @@
+/// Mode of address conversion. Refer tor [arti_client::TorAddr](https://docs.rs/arti-client/latest/arti_client/struct.TorAddr.html) for details.
+#[derive(Debug, Clone, Copy, Hash, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub enum AddressConversion {
+    /// uses only dns for address resolution (default)
```
```suggestion
    /// Uses only dns for address resolution (default).
```

For the sake of consistency, I would prefer comments to (a) be full sentences and (b) start with a capital letter.

mxinden reviewed (commented):
Thanks for the work here!

How about reaching out to the Arti maintainers, asking whether they would be so kind and give this a review as well? @umgefahren are you in touch with them already? I have been in touch with them in the past, thus happy to do the reach out.

umgefahren committed:
Apply suggestions from code review

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>
Co-authored-by: Max Inden <mail@max-inden.de>

umgefahren committed:
Added license headers and added to README.md

umgefahren committed:
Merge branch 'master' into add-onion

umgefahren committed:
Merge branch 'master' into add-onion

umgefahren committed:
addressed todos by @thomaseizinger

umgefahren commented on CHANGELOG.md:
```diff
@@ -48,6 +49,9 @@
 
+- New `libp2p-onion` crate
+  ([PR 2899](https://github.com/libp2p/rust-libp2p/pull/2899)).
+
```
I added a comment to the changelog for 0.49.0. Although I'm not actually sure this pull request will be merged in time and it's not mentioned in #2931

thomaseizinger committed:
Remove `cargo-deny` exception

The licenses are fine, we don't need an exception. We just have to
_clarify_ to `cargo deny`, what license it is.

thomaseizinger committed:
Remove unnecessary empty lines

thomaseizinger committed:
Merge `use` statements into one block

thomaseizinger committed:
Make use of link in changelog

thomaseizinger committed:
Consistently reference `Pin` with FQP to avoid import

thomaseizinger committed:
Update `arti` to latest version

thomaseizinger committed:
Sort dependencies alphabetically

thomaseizinger committed:
Use link reference in CHANGELOG.md

thomaseizinger committed:
Fix compile errors

thomaseizinger committed:
Merge branch 'master' into add-onion

thomaseizinger reviewed (approved):
This is ready to be merged!
Thank you @umgefahren :)

@mxinden Please go ahead with merging if it is also ready from your end.

umgefahren committed:
Merge branch 'master' into add-onion

umgefahren committed:
Merge branch 'master' into add-onion

umgefahren committed:
Really enable all features in full

umgefahren committed:
Merge branch 'master' into add-onion

umgefahren committed:
Merge branch 'master' into add-onion

umgefahren committed:
Adjusted example to match the ping example

thomaseizinger committed:
Fix formatting

umgefahren committed:
Merge branch 'master' into add-onion

thomaseizinger commented on CHANGELOG.md:
```diff
@@ -48,6 +49,8 @@
   new `full` feature which activates all features. See [PR 2918].
 
+
+- New `libp2p-onion` crate. See [PR 2899].
```
This needs to go into a new section of the changelog :)

umgefahren replied:
I moved it, I thought it would make it into the 0.49.0 release. Didn't expect the delays.

umgefahren committed:
Merge branch 'master' into add-onion

umgefahren committed:
Changed CHANGELOG.md entry

thomaseizinger commented on Cargo.toml:
```diff
@@ -49,6 +49,10 @@ full = [
+    "onion-async-std",
+    "onion-tokio",
+    "onion-native-tls",
+    "onion-rustls",
```
Please keep this list alphabetically sorted!

thomaseizinger commented on Cargo.toml:
```diff
@@ -85,7 +89,11 @@ secp256k1 = ["libp2p-core/secp256k1"]
+onion-native-tls = ["dep:libp2p-onion", "libp2p-onion?/native-tls"]
+onion-rustls = ["dep:libp2p-onion", "libp2p-onion?/rustls"]
+tokio = ["libp2p-mdns?/tokio", "libp2p-tcp?/tokio", "libp2p-dns?/tokio", "libp2p-onion?/tokio"]
 async-std = ["libp2p-mdns?/async-io", "libp2p-tcp?/async-io", "libp2p-dns?/async-std"]
```
This feature should activate the `async-std` feature in `libp2p-onion`, same for the tokio one.

thomaseizinger commented on Cargo.toml:
```diff
@@ -85,7 +89,11 @@ secp256k1 = ["libp2p-core/secp256k1"]
 serde = ["libp2p-core/serde", "libp2p-kad?/serde", "libp2p-gossipsub?/serde"]
-tokio = ["libp2p-mdns?/tokio", "libp2p-tcp?/tokio", "libp2p-dns?/tokio"]
+onion-async-std = ["dep:libp2p-onion", "libp2p-onion?/async-std"]
+onion-tokio = ["dep:libp2p-onion", "libp2p-onion?/tokio"]
```
Please remove these in favor of a single `onion` one and forwarding of the tokio and async-std ones in the `tokio` and `async-std` feature.

thomaseizinger commented on CHANGELOG.md:
```diff
@@ -43,11 +44,18 @@
 - Remove default features. You need to enable required features explicitly now. As a quick workaround, you may want to use the
   new `full` feature which activates all features. See [PR 2918].
 
+
```
```suggestion
```

thomaseizinger reviewed (approved):
LGTM!

A few minor changes please.
@mxinden are you happy to merge this?

umgefahren committed:
Update CHANGELOG.md

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

umgefahren committed:
Updated Cargo.toml responding to the requested changes

umgefahren committed:
Corrected error in the  feature flag

umgefahren committed:
Fixed issues with featues

thomaseizinger committed:
Alphabetically sort feature list

thomaseizinger committed:
Alphabetically sort dependency list

mxinden commented on transports/onion/Cargo.toml:
```diff
@@ -0,0 +1,30 @@
+[package]
+name = "libp2p-onion"
+version = "0.1.0"
```
```suggestion
version = "0.1.0-alpha"
```

Like with https://github.com/libp2p/rust-libp2p/pull/2622 and https://github.com/libp2p/rust-libp2p/pull/2289 I suggest releasing this new crate as an `alpha` for now. Objections?

umgefahren replied:
Sounds good to me.

mxinden replied:
:exploding_head: wow that was fast. Thanks @umgefahren.

mxinden reviewed (approved):
Merging here sounds good to me.

I reached out to @willscott on the licensing matters. He has no objections. I will reach out to our legal team as well, just in case. I don't think we need to block on this.

I suggest we wait for https://github.com/libp2p/test-plans/issues/59 and merge with a green CI though.

umgefahren committed:
Update transports/onion/Cargo.toml

Co-authored-by: Max Inden <mail@max-inden.de>

umgefahren committed:
Merge branch 'master' into add-onion

umgefahren committed:
Changed to alpha version.

umgefahren committed:
Merge branch 'master' into add-onion

mxinden committed:
Merge branch 'master' into add-onion

thomaseizinger committed:
Merge branch 'master' into add-onion

umgefahren committed:
Fixed issue with Cargo.toml

umgefahren committed:
Merge master into add-onion

umgefahren committed:
Added type alias for DataStream and removed custom doc(cfg( in order to comply to the new doc_auto_cfg policy

umgefahren committed:
Merge branch 'master' into add-onion

umgefahren committed:
Merge branch 'master' into add-onion

umgefahren committed:
Merge branch 'master' into add-onion

umgefahren committed:
Made compileable again
